From 529cb4df6a33a762b2f08ce42aafe19d727da44e Mon Sep 17 00:00:00 2001
From: Karthik T S <quic_kartikts@quicinc.com>
Date: Wed, 25 Oct 2023 20:48:04 +0530
Subject: [PATCH 334/500] l2tp: Add support for l2tpv3

1. Add "IFF_EXT_ETH_L2TPV3" flag in netdev to identify l2tp ethernet dev type.
2. IFF_EXT_ETH_L2TPV3 is set in "priv_flags_ext" field of netdev structure during netdevice intiialization.
3. Update the skb->skb->iif to copy interface index for all packets sent out on l2tpv3  ethernetinterface

Change-Id: I6ea7acae47fae8eac30f45bd7dbeb72e907df469
Signed-off-by: Karthik T S <quic_kartikts@quicinc.com>
---
 include/linux/netdevice.h | 2 ++
 net/l2tp/l2tp_eth.c       | 3 +++
 2 files changed, 5 insertions(+)

diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index b6fb8ed49b6a..57a73285cc71 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -1750,6 +1750,7 @@ enum netdev_priv_flags {
  * @IFF_EXT_GRE_V6_TAP: device is a GRE IPv6 TAP device
  * @IFF_EXT_IFB: device is an IFB device
  * @IFF_EXT_MAPT: device is an MAPT device
+ * @IFF_EXT_L2TPV3: device is a L2TPV3 Ethernet device
  */
 enum netdev_priv_flags_ext {
 	IFF_EXT_TUN_TAP			= 1<<0,
@@ -1761,6 +1762,7 @@ enum netdev_priv_flags_ext {
 	IFF_EXT_IFB				= 1<<6,
 	IFF_EXT_MAPT			= 1<<7,
 	IFF_EXT_HW_NO_OFFLOAD		= 1<<8,
+	IFF_EXT_ETH_L2TPV3		= 1<<9,
 };
 
 #define IFF_802_1Q_VLAN			IFF_802_1Q_VLAN
diff --git a/net/l2tp/l2tp_eth.c b/net/l2tp/l2tp_eth.c
index f2ae03c40473..97b827319963 100644
--- a/net/l2tp/l2tp_eth.c
+++ b/net/l2tp/l2tp_eth.c
@@ -76,6 +76,8 @@ static netdev_tx_t l2tp_eth_dev_xmit(struct sk_buff *skb, struct net_device *dev
 	struct l2tp_eth *priv = netdev_priv(dev);
 	struct l2tp_session *session = priv->session;
 	unsigned int len = skb->len;
+
+	skb->skb_iif = dev->ifindex;
 	int ret = l2tp_xmit_skb(session, skb);
 
 	if (likely(ret == NET_XMIT_SUCCESS)) {
@@ -117,6 +119,7 @@ static void l2tp_eth_dev_setup(struct net_device *dev)
 	SET_NETDEV_DEVTYPE(dev, &l2tpeth_type);
 	ether_setup(dev);
 	dev->priv_flags		&= ~IFF_TX_SKB_SHARING;
+	dev->priv_flags_ext     |= IFF_EXT_ETH_L2TPV3;
 	dev->features		|= NETIF_F_LLTX;
 	dev->netdev_ops		= &l2tp_eth_netdev_ops;
 	dev->needs_free_netdev	= true;
-- 
2.34.1

